### ISSUE TITLE
Native WebSocket support

### ISSUE THREAD

**Issue by net** [SergioBenitez/Rocket#90]
I think Rocket would benefit heavily from native WebSocket support. Even better would be something similar to [Phoenix's channels](http://www.phoenixframework.org/docs/channels).

It's great to see such a comprehensive web framework for Rust coming together!

**Comment by jsen-** [comment]
@net did you manage to combine `rocket` with eg. [websocket](https://crates.io/crates/websocket) on the same `host:port`?
I'm trying to `upgrade` connection to `ws`, but no luck so far :(

**Comment by SergioBenitez** [comment]
This is indeed planned.

**Comment by wagenet** [comment]
@SergioBenitez got any suggestions for how one could manually add web socket support?

**Comment by SergioBenitez** [comment]
@wagenet Do you mean to Rocket or to your web application? The former is subtle is it requires a lot of design work, but the latter should be straightforward. You should be able to use the [`websocket`](https://github.com/cyderize/rust-websocket) or [`ws-rs`](https://github.com/housleyjk/ws-rs) crates as they are intended. You won't be able to use the same port for both servers, of course. And, you'll need to spawn one of the two servers in a separate thread so that one server doesn't block the other one from starting. Should you choose to do this, your `main` function would look something like:

```rust
fn main() {
    thread::spawn(|| {
        listen("127.0.0.1:3012", |out| {
            move |msg| {
                out.send(msg)
           }
        })
    })

    rocket.ignite()..more()..launch()
}
```

**Comment by wagenet** [comment]
@SergioBenitez thanks :) I ended up discovering this sort of approach myself.

**Comment by jhpratt** [comment]
@SergioBenitez is this waiting on a future version of Rocket that migrates to Actix or some other async framework?

**Comment by sparky8251** [comment]
Is there no way to add web socket support on the same port as HTTP?

If not, this is quite problematic for me.

**Comment by jhpratt** [comment]
@sparky8251 Not currently, at least easily.

**Comment by sparky8251** [comment]
Well, that sucks. Liked the readability of Rocket but without the ability to have websockets on the same port Rocket is useless to me.

Actix just doesnt have the same readability after all... If you happen to know a way to do web sockets on the same port, even if its not clean, I'd love to hear it. If I end up porting to Actix I won't be returning to Rocket but if I can make the web socket stuff work, hopefully I can stick around until native support lands.

**Comment by jebrosen** [comment]
@sparky8251 nginx ("Since version 1.3.13") should be able to forward to a second server on a per-location basis. The second server could even be another listening port connected to the same binary that's serving the rocket HTTP server (e.g. via https://github.com/SergioBenitez/Rocket/issues/90#issuecomment-304725055).

I've made an example (completely untested so far, sorry) based on [the nginx example for WebSocket](https://nginx.org/en/docs/http/websocket.html):

```
http {
    map $http_upgrade $connection_upgrade {
        default upgrade;
        ''      close;
    }

    server {
        listen 80 default server;
        server_name example.com;

        proxy_pass http://localhost:8000;
        proxy_http_version 1.1;
        proxy_set_header Upgrade $http_upgrade;
        proxy_set_header Connection $connection_upgrade;

        location /websocket/ {
            proxy_pass http://localhost:8080;
            proxy_http_version 1.1;
            proxy_set_header Upgrade $http_upgrade;
            proxy_set_header Connection $connection_upgrade;
        }
    }
```

I presume something similar can be done with Apache as well.

**Comment by sparky8251** [comment]
Unfortunately, using nginx is a non-answer for my use case. I'd like to not need a 3rd party program to make this work.

**Comment by VictorKoenders** [comment]
Hyper 0.12 supports websockets: https://github.com/hyperium/hyper/blob/master/examples/upgrades.rs

Rocket is still on hyper 0.10 though, so we'll need to upgrade that dependency first

**Comment by fenhl** [comment]
With #1008 merged, would it make sense to start implementing websocket support on the async branch?

**Comment by jebrosen** [comment]
I think we would need to see some progress on #1066 first - which reminds me, I need to write a bit more about that. And WebSocket support is also big enough that I'd like to see a hypothetical API design before going too far into implementation.

**Comment by jhpratt** [comment]
WebSockets will certainly be one of the larger issues. I'm actually looking into the two main websocket crates (`ws` and `websocket`). `ws` is, frankly, a bit of a mess, and will be extremely difficult to upgrade to asynchronous. `websocket` already has support, though it's running on tokio 0.1 and hyper 0.10. I've just asked on an issue there to see what sort of changes will be accepted.

While this wouldn't directly add websocket support to Rocket, it would allow for an asynchronous server that could be spawned on the same runtime.

**Comment by jebrosen** [comment]
Actually, I don't think #1066 is necessary for this after all - it would help with SSE, but not websockets. Instead, websockets would likely have to be handled at the hyper-service or connection level with an API such as https://docs.rs/websocket/0.23.0/websocket/server/upgrade/async/trait.IntoWs.html.

**Comment by schrieveslaach** [comment]
@jebrosen, based on your `async` branch I created a branch with rudimentary websocket support and I would like to contribute this new feature to rocket. 

The first draft of the API looks likes this: 

```rust
let (tx, rx) = channel();

std::thread::spawn(move || {
    let duration = std::time::Duration::from_secs(1);
    loop {
        println!("Sending message");
        tx.unbounded_send(Message{}).unwrap();
        std::thread::sleep(duration);
    }
});

let _ = rocket::ignite().receivers(vec![rx]).mount("/", routes![hello]).launch();
```

@SergioBenitez, @jebrosen, what do you think about this first implementation. 

BTW: [here is the branch](https://github.com/schrieveslaach/Rocket/tree/websockets), implementing the behavior.

**Comment by jebrosen** [comment]
My vague idea for a websocket API was to be able to do something like the following:

```rust
#[websocket("/chat/<room>")]
async fn chat(room: String, conn: WebSocket) {
    // let message = conn.recv().await;
    // conn.send(message).await;
}

fn main() {
    rocket::ignite().websocket(websocket![chat]);
}
```

That is, have it work similarly and in parallel to `catcher`s and `route`s. It should also have a lower-level API like `Handler`. I'm not sure if we should support only Websocket or expose more flexibility in terms of other protocols via `Upgrade`.

As far as I can tell your current implementation takes any number of receivers, and every message that comes on any receiver is sent to all connected clients - and that pattern is too limiting.

**Comment by schrieveslaach** [comment]
Okay, that looks much nicer, but I have one question. If I understand your example correctly, wouldn't this require that Rocket run the `chat` function every `x` milliseconds? 

> As far as I can tell your current implementation takes any number of receivers, and every message that comes on any receiver is sent to all connected clients - and that pattern is too limiting.

This is not intended to be the final result of a future PR. It was just a very basic implementation to get it running. 

Currently, I'm wondering if we could combine our API designs:

```rust
#[websocket("/chat/<room>")]
async fn chat(room: String, msg: Message) -> Json<Value> {
    // Will be invoced when rx emits a message and here you could 
   // transform the message into a certain paylod.
}

fn main() {
    let (tx, rx) = channel();

    std::thread::spawn(move || {
        let duration = std::time::Duration::from_secs(1);
        loop {
            println!("Sending message");
            // Message could be a trait and here you could create a 
            // text message (similar to websockets crate API)
            tx.unbounded_send(Message{}).unwrap();
            std::thread::sleep(duration);
        }
    });

    rocket::ignite().websocket(vec![websocket!(chat, rx)]);
}
```

**Comment by fenhl** [comment]
This API still seems very limited, since it wouldn't be able to send messages on its own, or respond to a message with multiple messages. There's also a lot of boilerplate in `main`, which seems like it would be unnecessary. A websocket handler that continues to handle the websocket connection for its entire lifetime, as in @jebrosen's example, would be ideal.

**Comment by schrieveslaach** [comment]
I'm a bit confused how you would receive message from a background process (imaging that you are receiving Kafka messages from another service in the background and want to update the UI through a websocket connection). Could you give me an example API?

**Comment by fenhl** [comment]
It's hard to give a code example without knowing what the message producer API looks like, but it would probably involve [`State`](https://docs.rs/rocket/0.4.2/rocket/struct.State.html).

**Comment by jebrosen** [comment]
I previously wrote a chat room example with async rocket and SSE, where the "receiver" endpoint looks like this: https://git.jebrosen.com/jeb/rocket-rooms/src/commit/381f93daaf17db1a430760f536ceb9699c33b2ea/src/main.rs#L34

The websocket equivalent could end up looking something like this:

```rust
#[websocket("/room/<room>")]
async fn chat_room(room: String, rooms: State<'_, Rooms<Message>>, socket: WebSocket) {
    // Subscribe to the room. 'subscription' is a Stream of Messages.
    let mut subscription = rooms.subscribe(&room).await;

    // WebSockets are full duplex, and we want to read/write independently
    let (tx, rx) = socket.split();

    loop {
        // TODO: I have no idea if select! actually works like this
        futures::select! {
            // New message from another client: send it to this one
            server_msg = subscription => tx.send(server_msg).await,

            // New message from this client: broadcast it to the others
            client_msg = rx => rooms.send(&room, client_msg).await,

            // TODO: error handling, shutdown, etc.
        }
    }
}
```

The main thing this showcases is that neither sending nor receiving of messages is directly controlled by rocket, so the handler can receive from any source.

**Comment by schrieveslaach** [comment]
That example is the example I was looking for in order to understand what should be accomplished. At the weekend I'll have a look into it.

**Comment by nour-dev** [comment]
any news about Websocket support with rocket ?

**Comment by DrYSG** [comment]
@jebrosen 

Can you say anything to compare:

1. The work you did
2. The work shown at: https://www.steadylearner.com/blog/read/How-to-start-Rust-Chat-App
3. What are the cons/pros of native rocket support vs. the two above hybrid approaches?

**Comment by jebrosen** [comment]
@DrYSG 

1. I haven't done any actual work on websockets, other than provide an example for nginx that I have not tested myself. The purpose of that example is to map servers on two ports to look like they are on the same port.
2. That example starts a websocket and a webserver app in the same process listening on two ports. It does not look like the two share data in any way; if needed, the two could communicate through some shared `Arc<Mutex<State>>` or channel(s).
3. Native rocket support would mean that both HTTP and websocket would truly be handled by the same server listening on the same port, and it would simplify sharing state: the websocket routes could access the same managed state through the same mechanism as non-websocket routes.
4. Warp's websocket support is the same kind as number 3 - the same server handles both plain and websocket requests.

**Comment by rustrust** [comment]
see https://github.com/SergioBenitez/Rocket/issues/1238

**Comment by schrieveslaach** [comment]
I worked on native websocket support for Rocket (see my comments above). However, I did not had time to continue because I think it requires large extensions and maybe refactorings of Rocket (maybe a second kind of handler, as I tried to use [here](https://github.com/schrieveslaach/Rocket/commit/1b5bf1da1fc81c29cf972cf19b5b9cdf8fe6e112#diff-a9ea99dbf2f3bc38a7d7e4c36a99a288R297-R312)). Maybe, we could discuss this here?

@rustrust, using a test suite would very beneficial. :+1:

**Comment by Voodlaz** [comment]
I think that I'm going to work with websockets.

**Comment by Yuhanun** [comment]
Is there any update on when this is going to release? In a dire need for WebSocket endpoints.

**Comment by howard0su** [comment]
I am thinking something different model as @jebrosen.

We should still use the current routing logic but add a special responder, which will issue the upgrade response. In this new responder, the client application need supply async entry point which accept a websocket. The websocket logic runs as an async function and give the client application flexibility. This also enables one handler is able to reject the wrong websocket connection requests with the wrong parameters. Request Guard still works.

**Comment by Voodlaz** [comment]
I've got time to work with this. What I can do now for implementing websockets?

**Comment by Voodlaz** [comment]
@jebrosen @SergioBenitez ?

**Comment by jebrosen** [comment]
> I've got time to work on implementing websockets. What I can do now for implementing websockets?

Since WebSocket (and HTTP upgrade in general) is a pretty significant API addition that also needs some internal changes, I expect several false starts or reworks of the design/implementation. I don't have any particular suggestions for a first step, but there are a lot of different areas to consider.


Here's a very rough idea of what Rocket will need to do to handle an incoming WebSocket connection, which could be expanded on to make a complete design:

1. Look for and validate the [client's WebSocket opening handshake](https://tools.ietf.org/html/rfc6455#section-4.2.1).
2. In the case of a valid handshake, [do an HTTP upgrade](https://docs.rs/hyper/0.13.7/hyper/upgrade/index.html), [completing the server's WebSocket opening handshake](https://tools.ietf.org/html/rfc6455#section-4.2.2) instead of processing it like a normal HTTP request.
3. Wrap the I/O in something that speaks the WebSocket protocol, such as `tokio-tungstenite` (or another crate).
4. Run the appropriate route handler, passing it a `tokio_tungstenite::WebSocketStream` (or whichever type).

This still leaves some significant open questions:
* We likely want a new route attribute, like `#[websocket]`, and the changes to `Route` that go with it.
* Should Rocket expose the HTTP upgrade mechanism directly, or should it implement WebSocket only, or should both be part of the API? My preference is ultimately "both", but we might have one be a public/stable API for a while before we expose the other.
* The current request handling logic takes the incoming hyper `Body`, including the `.on_upgrade()` function needed for step 2 above, and encapsulates it into `Data`. We'll need to expose `on_upgrade` through `Data` and/or not use `Data` for routes that are an HTTP upgrade.
* Do `on_request` fairings apply to websocket/upgrade routes? It seems likey they should, but they take an `&Data` which might be a problem.

I think at this stage the biggest unknown is how much of the existing code needs to change to accommodate websockets/upgrade, and how many public APIs "have to" break. I haven't looked at it for a while, but the branch linked in <https://github.com/SergioBenitez/Rocket/issues/90#issuecomment-541434925> is a good start as far as answering those kinds of questions.

**Comment by schrieveslaach** [comment]
@Voodlaz I tried to integrate websocket support but I had no time to continue. I was stuck at the point where I wanted to integrate websockets into the codegen of Rocket. Unfortunately, I couldn't find a good way to integrate the [second handler](https://github.com/schrieveslaach/Rocket/blob/1b5bf1da1fc81c29cf972cf19b5b9cdf8fe6e112/core/lib/src/handler.rs#L332) into tho codegen and I wasn't sure if I follow the right approach.

**Comment by markazmierczak** [comment]
I have created a [pull-request](https://github.com/SergioBenitez/Rocket/pull/1441) with minimal support for WebSockets based on tungstenite crate and I am seeking for comments.
It would be very helpful to know if this path is desired.

Personally I am with @howard0su and I think `#[websocket(..)]` macro has drawbacks.
E.g. there is no way to validate if `room` exists before accepting WebSocket connection.
IMHO returning _404 Not Found_ is what is desired in case a room does not exist.

Nevertheless, adding macro later is still the option and it is on my list.
The very next step I want to take is to write Chat example with current approach.
But before I dive into that I am looking forward to hear what you think about it.

**Comment by ducharmemp** [comment]
So I played around with the above PR, big fan of how easy tungstenite is to use. Some thoughts:

Rocket should probably handle the async task spawning and maybe teardown internally. With the approach in the PR I *think* a malicious user can connect to the server as many times as they want and hit some OS limits on open fds (haven't tested so kind of hand waving). 

The approach in the PR also doesn't currently allow for sharing/grabbing the managed state from rocket or holding a ref from the request's state. I think doing that part is easy enough (adding a state field to some struct and copying the request's state field), the tricky bit is allowing the user to grab said state ad hoc or based on some factors. This is necessary if the user wants to use rocket to manage their databases, like I do. It's even trickier when considering how to make this feel like normal rocket.

What I *think* might be a correct approach is something a bit more evented/callback based. Maybe something like

```rust

#webhook("/ws/chat")
async fn chat(ws: Websocket) {
    let (tx, rx) = unbounded_channel();
    ws.on_message(|conn: &PgConnection| async move {
       ... 
       tx.send("foo").await;
   });

   rx.for_each(ws.on_send(|conn: &PgConnection| {....});
}
```

No idea what that function should return since I'm on my phone and throwing around these ideas, but I think that basically it should not return until the websocket should get closed. The websocket macro should be a basic wrapper around the get macro and should set up the boilerplate for actually initializing the websocket and completing the handshake, and then spawning off to the decorated function. That spawn should take into account the number of already spawned clients and not spawn any more if there are too many (which would mean a 5xx error?). It would also be able to construct a Websocket instance which would need to hold onto the managed state from rocket until it gets dropped.

Apologies for anything a bit too out of whack or hard to parse, I'm on my phone atm and had some thoughts I wanted to jot down.

**Comment by ducharmemp** [comment]
I did some work based on #1441 (have to amend the message so there's no more spam here), current ref here https://github.com/ducharmemp/Rocket/commit/30f85466c11d2297d002af9f11dedeb708a79329

When I did some testing, there didn't seem to be much of a downside to delaying the on_upgrade call until after the response is sent, but I think I'll need to talk to someone more experienced with hyper and HTTP. Note that the above mentioned branch is very WIP, this still needs

a) Local testing capabilities
b) A macro to hide the Box::pin awfulness that exists in the example handler
c) Cleanup of the handlers in general to not be a copy-pasted and tweaked version of the `lib/handlers.rs` traits/types
d) Some nice way of fetching state from the global rocket instance that's a bit more hidden. 
e) General cleanups and documentation

I think having this surfaced as an initial GET request and a return of something that initiates an upgrade based on the response is important, this means that for a path of `#[get("/foo/bar/<id>")]` we can first validate if the id actually exists in a db and return the appropriate 4xx response code, killing the attempted connection before it starts.

**Comment by adhesivee** [comment]
I want to share another idea how to approach this.
What I see with other framworks is that websocket sessions are handled by the framework itself. 

For example, for this you could create an enum for this:
```rust
pub enum WebSocket {
    Open(Session),
    TextMessage(Session, Message),
    Close(Session)
}
```

Then we could define it similar as what have been proposed here earlier:
```rust
#[websocket("/ws/chat")]
fn chat(ws: Websocket) {
  match ws {
        WebSocket::Open(_) => {}
        WebSocket::TextMessage(_, _) => {}
        WebSocket::Close(_) => {}
    }
}
```
The idea with above example is that after each action, the methods returns and will not wait.

I think it will solve/help with some issues described above:
- because of exhaustive match, users see directly all capabilities
- using states could be fixed, each time an action occurs, it will enter the method "fresh", it can then also fetch State how it is currently doing
- session in above example could hold a value to underlying capabilities of websocket channel for more control (if possible)

One note, I am not that experienced in Rust. I don't know if this is "idiomatic" Rust, or that it may "violate" some rules. 
But maybe a different view might help here.

**Comment by the10thWiz** [comment]
@SergioBenitez I would like to work on this issue, and I have some ideas on how to implement this feature.

I'm imagining an API that looks something like this: 

```rust
#[get("/rooms/join/<name>")]
async fn join_room(name: String, app: State<'_, AppData>) -> Websocket {
    Websocket::create(|ws| async {
        ws.send("example_message").await?;
        while let Some(message) = ws.recv().await {
            ws.send(format!("Recieved: {}", message)).await?;
        }
        Ok(())
    })
}
```

This example sends an initial message, and responds to any message with a similar message, essetially just an echo. To set options, extensions and other important parts of the websocket connection, they would be configured on the websocket object, before returning it. If the websocket connection might fail because the endpoint might not exist, i.e. attempting to connect to a room that doesn't exist, you would return a `Result<Websocket, Status>`, similar to a normal response.

The `Websocket::create` function signature would look something like:

```rust
impl Websocket {
    fn create<F, T>(ws_task: F) -> Self where 
        F: FnOnce(WebsocketChannel) -> T,
        T: Future<Output = io::Result<()>> + Send + 'static;
}
```

The `ws_task` parameter is a function that accepts a `WebsocketChannel` object, and returns a future, which returns a `io::Result<()>`. The `WebsocketChannel` implements methods for sending and recieving messages sent on the channel, and closes the channel when dropped. It would be responsible for the `hyper::upgrade::Upgraded` object that holds the underlying streams. Rocket would be responsible for calling and managing the lifetime of `ws_task`, which hopefully shouldn't be too hard.

At this point, there would be no way to limit the number of concurrent open connections, beyond implementing it manually via a request guard or something similar. I think this is acceptable, and if such an implementation is desired by a large enough group of people, it could be added to `rocket_contrib`.

The biggest issue remaining is the message data types. The RFC defines this in terms of Frames, but since a single message can be spread across multiple frames, it doesn't make sense to expose this detail to the user, rather we should provide a Message, which can be one or more frames. The websocket RFC defines Text (Utf-8) and Binary data types for a message, and I think a generic `Message` struct, which holds the raw payload as a `Vec<u8>`, along with the associated datatype sent by the client is the simplest awnser. A number of other objects, such as `String` could implment `From<Message>`, and the send/recv methods could use that to make converting a message into a useable form much easier. The RFC also defines a set of control frames, which almost certainly don't need to be exposed to the user. Close can be handled internally, and sent when appropriate, while Ping and Pong messages should be reasonably easily handled. It might make sense to provide a `ping` method on `WebsocketChannel`, which sends a Ping, and waits for the associated Pong.

As for the internal API, I imagine that it might be simplest to expose the hyper upgrade mechanisms to a response, should they be desired. I imagine this would be an addition function on the `Responder` trait, with a default implementation. Essentially, rather than just calling `respond_to`, it would first call `upgrade`, and if that returns None, it would then call `respond_to`. The default implementation for `upgrade` just fails through to the respond_to method, which should make it backwards compatible. This would also allow custom http upgrades.

I imagine that the http upgrade mechanisms should be added to Rocket itself, but the Websocket implementation can be added to Rocket behind a feature flag, or to Rocket Contrib. Which would you prefer?

Please let me know if this sounds like it could work. I am interested in implementing this into Rocket, and I don't want to spend too long working on it before I know it's good.

**Comment by tvallotton** [comment]
> My vague idea for a websocket API was to be able to do something like the following:
> 
> ```rust
> #[websocket("/chat/<room>")]
> async fn chat(room: String, conn: WebSocket) {
>     // let message = conn.recv().await;
>     // conn.send(message).await;
> }
> 
> fn main() {
>     rocket::ignite().websocket(websocket![chat]);
> }
> ```
> 
> That is, have it work similarly and in parallel to `catcher`s and `route`s. It should also have a lower-level API like `Handler`. I'm not sure if we should support only Websocket or expose more flexibility in terms of other protocols via `Upgrade`.
> 
> As far as I can tell your current implementation takes any number of receivers, and every message that comes on any receiver is sent to all connected clients - and that pattern is too limiting.


Just is a suggestion, I was thinking on an API like this:
```rust
#[websocket("/chat/<room>")]
async fn chat(messages: SomeType) {
    for message in messages {
        yield message;
    }
}
```
I suppose the yield statement could later be replaced with an await by the macro, so it could compile on stable.

**Comment by tvallotton** [comment]
> My vague idea for a websocket API was to be able to do something like the following:
> 
> ```rust
> #[websocket("/chat/<room>")]
> async fn chat(room: String, conn: WebSocket) {
>     // let message = conn.recv().await;
>     // conn.send(message).await;
> }
> 
> fn main() {
>     rocket::ignite().websocket(websocket![chat]);
> }
> ```
> 
> That is, have it work similarly and in parallel to `catcher`s and `route`s. It should also have a lower-level API like `Handler`. I'm not sure if we should support only Websocket or expose more flexibility in terms of other protocols via `Upgrade`.
> 
> As far as I can tell your current implementation takes any number of receivers, and every message that comes on any receiver is sent to all connected clients - and that pattern is too limiting.


Just is a suggestion, I was thinking on an API like this:
```rust
#[websocket("/chat/<room>")]
async fn chat(messages: SomeType) {
    for message in messages {
        yield message;
    }
}
```
I suppose the yield statement could later be replaced with an await by the macro, so it could compile on stable.

**Comment by the10thWiz** [comment]
Actually, an interesting idea is something more like 

```rust
#[websocket("/echo")]
async fn echo(message: Message, ws: Websocket, channel: WebsocketChannel) {
    ws.send(message).await;
}
```

The `Message` would contain the message sent (the user would have to extract the specified room if they need it),
and the `Websocket` would contain a handle to the websocket that sent the message. The `WebsocketChannel` has a refernce to the larger websocket pool, and would have methods that allow the user to send to a whole room, or to any other open websocket. I don't think we actually need a special `websocket![]` macro, since we should be able to make it just work with the normal `routes![]` macro.

The function would be called on every incomming message from a client connected to the endpoint. Since Websockets aren't quite like HTTP, I don't think it makes sense to have a return value, but allowing the user to return a Result of some type (probably whatever send returns) would allow the user to add `?`.

**Comment by SergioBenitez** [comment]
For those watching: design discussions are taking place on Matrix. Here is my proposed API for websockets, by example:

```rust
//! A multiroom chat application with a special `echo` room which echoes its
//! messages to `lobby`.

// `&Channel`, `&Broker` are channel guards. `&Broker` is also a request guard.
//
// A `Channel` is:
//   1) A bidrectional channel between two endpoints: client and server.
//   2) Linked to a topic URI, for broadcasting to that topic.
//
// A `Broker` knows about all channels.

type UserList<'r> = State<'r, ConcurrentHashSet<String>>;

enum Event<'r> {
    Join(&'r str),
    Message(&'r str),
    Leave(&'r str),
}

/// A new client has connected to a room. Inform the room of this, say hello to
/// the client with a "join" message.
#[join("/<_>", data = "<username>")]
fn join(channel: &Channel, username: &str, users: UserList<'_>) {
    if users.contains_or_add(username) {
        channel.send(Event::Leave("username taken")).await;
        return channel.close();
    }

    channel.local_cache(|| username.to_string());
    channel.send(Event::Join("hello!")).await;
    channel.broadcast(Event::Join(username)).await;
}

/// A message sent to a room that's not `echo`.
#[message("/<_>", data = "<msg>")]
fn room(channel: &Channel, msg: &str) {
    channel.broadcast(Event::Message(msg)).await;
}

/// A message sent to `/echo`: send it to `/echo` and `/lobby` concurrently.
#[message("/echo", data = "<msg>")]
fn echo(broker: &Broker, channel: &Channel, msg: &str) {
    let lobby_send = channel!(broker, room("/lobby")).broadcast(Event::Message(msg));
    let room_send = channel.broadcast(Event::Message(msg));
    join!(lobby_send, room_send).await;
}

/// A user has left a room. Let the room know. `Username` retrieves the username
/// stored in `join` in `channel-local cache` via an elided channel guard impl.
#[leave("/<_>")]
fn leave(channel: &Channel, username: &Username, users: UserList<'_>) {
    users.remove(username);
    channel.broadcast(Event::Leave(username)).await;
}
```

**Comment by iMplode-nZ** [comment]
So I guess there isn't any working prototype for this and I should just use the preexisting websocket thing mentioned above?

**Comment by iMplode-nZ** [comment]
Also maybe add `tungstenite` to the possible options?

**Comment by ponyatov** [comment]
Can SocketIO.js on the client side be used as a fallback for Rocket in longpoll mode?
Maybe it can be a more light variant requires much lesser implementation efforts, or using already existing infrastructure.

**Comment by jebrosen** [comment]
> Maybe it can be a more light variant requires much lesser implementation efforts, or using already existing infrastructure.

No; in fact this would probably require even more implementation efforts: Socket.IO uses its own protocol on top of long-polling and/or WebSockets, and the few rust implementations of socket.io servers I found appear to be ready or suitable for "plugging in" to Rocket at the moment.

**Comment by the10thWiz** [comment]
> So I guess there isn't any working prototype for this and I should just use the preexisting websocket thing mentioned above?

If you are interested, there is a pull request I've been working on to integrate websocket support directly into Rocket itself. However, I would caution against using it in production yet, since it's not quite ready yet.

There are several option questions related to websocket support, and some serious testing need to happen. Since it sounds like you are implementing a service that uses websockets, how do you handle user authentication? I've implemented a solution that takes advantage of temporary URLs, but I'm curious to hear what your solution is.

**Comment by iMplode-nZ** [comment]
Well for emulating WS via EventStream, I just send out a unique token and use that as the data field and use a custom eventstream impl on the client that supports payloads.

**Comment by Kiiyya** [comment]
Looks like WebSockets are currently planned for the 0.8.0 milestone. Sounds it'll be a few years still, then?

**Comment by hf29h8sh321** [comment]
[This proposed API](https://github.com/SergioBenitez/Rocket/issues/90#issuecomment-831610067) is the best IMO, since it offers the greatest flexibility. A channel-based API may be too limiting for certain use cases.

**Comment by TimBoettcher** [comment]
Did the design discussions get anywhere? I have a feeling this is a highly anticipated feature, and I myself would appreciate it for my app, too, but it looks like it's not feasible to implement for the time being?

Another question: Does rocket somehow expose internal state? Because if I run a separate server for WS handling (eg. with [tokio-tungstenite](https://crates.io/crates/tokio-tungstenite)), I'd need to somehow get the internal state and possibly DB pool of the rocket app, I think.

**Comment by toxeus** [comment]
@TimBoettcher I recommend reading up the [discussion on the PR](https://github.com/SergioBenitez/Rocket/pull/1639#issuecomment-924899064) to get an idea about alternatives and why rocket cannot be used with tokio-tungstenite.

**Comment by tvallotton** [comment]
What I really like about [this proposal](https://github.com/SergioBenitez/Rocket/issues/90#issuecomment-831610067) is that is the one that is most consistent with Rockets current API. Macros and guards generally deal with request from the user, while the response is dealt by the return type . Here `Websocket` would be just an ordinary type that implements `Handler`.

**Comment by hf29h8sh321** [comment]
I've seen [this crate](https://crates.io/crates/hyper-tungstenite) which might provide a somewhat easy way to provide an implementation.

**Comment by leapofazzam123** [comment]
[This one](https://github.com/SergioBenitez/Rocket/issues/90#issuecomment-831610067) is definitely better than most proposals. The WebSocket request API isn't too complex I could easily write it. It's also very flexible than what most other proposal offers.

**Comment by Mai-Lapyst** [comment]
Worked the last few weeks on a solution similar to the one presented in [this comment](https://github.com/SergioBenitez/Rocket/issues/90#issuecomment-831610067), by adding an API to expose the hyper upgrade to rocket and implementing websockets with it. Have submitted an combined PR (#2466) for you guys to try it out.

There are some things missing for a complete websocket implementation such as subprotocols and extensions; but for now simple sending/recieving should suffice.

Example for the upgrade API:
```rust
struct TestSocket {}

#[crate::async_trait]
impl Upgrade<'static> for TestSocket {
  async fn start(&self, upgraded: crate::http::hyper::upgrade::Upgraded) {
    // can fully use the hyper::upgrade::Upgraded struct
  }
}

impl<'r, F> Responder<'r, 'r> for TestSocket {
  fn respond_to(self, req: &'r Request<'_>) -> response::Result<'r> {
    Response::build()
      .status(Status::SwitchingProtocols)
      .raw_header("Connection", "upgrade")
      .raw_header("Upgrade", "testsocket")
      .upgrade(Some(Box::new(self)))
      .ok()
  }
}
```

and one for the websockets:
```rust
use rocket::response::websocket::{WebsocketMessage, WebsocketChannel, CreateWebsocket, Websocket};

#[get("/ws")]
fn websock() -> Websocket![] {
  CreateWebsocket! {
    while let Some(msg) = ch.next().await {
      println!("handler {msg:?}");
      ch.send(msg);
    }
  }
}
```

**Comment by SergioBenitez** [comment]
Connection upgrading support just landed via https://github.com/SergioBenitez/Rocket/pull/2488 and https://github.com/SergioBenitez/Rocket/commit/d97c83d7e0989f4a11a026048392de7ead776a67! Thanks to @Mai-Lapyst for the push and idea. This means we can implement support for WebSockets outside of Rocket. The new [upgrade example](https://github.com/SergioBenitez/Rocket/tree/bd482081ad58b68fdcf2398d87ddda4d8e850964/examples/upgrade) does exactly this! Here's the `echo` server:

```rust
#[get("/echo")]
fn echo(ws: ws::WebSocket) -> ws::Channel {
    ws.channel(|mut stream| Box::pin(async move {
        while let Some(message) = stream.next().await {
            let _ = stream.send(message?).await;
        }

        Ok(())
    }))
}
```

I still believe we want an API like the one in https://github.com/SergioBenitez/Rocket/issues/90#issuecomment-841866637, but this brings us to parity with the echo system. Soon, I'll modify the example to make use of Rocket's async stream supports and macros, which will make the example closer to:

```rust
#[get("/echo")]
fn echo(ws: ws::WebSocket) -> ws::Channel {
    ws::channel! { ws =>
        for await message in ws {
            yield message;
        }
    }
}

```

**Comment by SergioBenitez** [comment]
The stream API is now implemented in the example!

https://github.com/SergioBenitez/Rocket/blob/2abddd923e36b49b01821f09f4dc82ee19d69dcc/examples/upgrade/src/main.rs#L19-L26

**Comment by SergioBenitez** [comment]
WebSocket support is now in the contrib `ws` crate. Full docs are available at [api.rocket.rs](https://api.rocket.rs/master/rocket_ws/index.html). Note that the crate is not yet released, so it cannot be used off of crates.io and must be used from the `git` repository until the next release. If you wish to test this, you'll also need to depend on `rocket` from GitHub (as well as `rocket_ws` from GitHub, of course).

As WebSocket support is now available from an officially supported crate, I believe we can close this issue. I would still like Rocket to eventually support a channels/group-based API as in https://github.com/SergioBenitez/Rocket/issues/90#issuecomment-841866637, but we can leave that for another issue.

Thank you all, for your patience. ❤️

**Comment by SUIBING112233** [comment]
Congratulations! This feature is really important and I'm really excited to see WebSocket merged into Rocket!

**Comment by HarikrishnanBalagopal** [comment]
> WebSocket support is now in the contrib `ws` crate. Full docs are available at [api.rocket.rs](https://api.rocket.rs/master/rocket_ws/index.html). Note that the crate is not yet released, so it cannot be used off of crates.io and must be used from the `git` repository until the next release. If you wish to test this, you'll also need to depend on `rocket` from GitHub (as well as `rocket_ws` from GitHub, of course).
> 
> As WebSocket support is now available from an officially supported crate, I believe we can close this issue. I would still like Rocket to eventually support a channels/group-based API as in [#90 (comment)](https://github.com/SergioBenitez/Rocket/issues/90#issuecomment-841866637), but we can leave that for another issue.
> 
> Thank you all, for your patience. ❤️

The docs seem to be wrong

The docs say to do
<img width="594" alt="image" src="https://github.com/SergioBenitez/Rocket/assets/20921177/cdff2000-6d7e-4d87-9168-9f16b4b89f22">

but the crate is not yet in crates.io so you need to do
https://github.com/dani-garcia/vaultwarden/blob/bc26bfa589c007da9b9be37e1172060f38a948b9/Cargo.toml#L60
```
rocket_ws = { git = 'https://github.com/SergioBenitez/Rocket', rev = "ce441b5f46fdf5cd99cb32b8b8638835e4c2a5fa" } # v0.5 branch
```

**Comment by SergioBenitez** [comment]
> > WebSocket support is now in the contrib `ws` crate. Full docs are available at [api.rocket.rs](https://api.rocket.rs/master/rocket_ws/index.html). Note that the crate is not yet released, so it cannot be used off of crates.io and must be used from the `git` repository until the next release. If you wish to test this, you'll also need to depend on `rocket` from GitHub (as well as `rocket_ws` from GitHub, of course).
> 
> > 
> 
> > As WebSocket support is now available from an officially supported crate, I believe we can close this issue. I would still like Rocket to eventually support a channels/group-based API as in [#90 (comment)](https://github.com/SergioBenitez/Rocket/issues/90#issuecomment-841866637), but we can leave that for another issue.
> 
> > 
> 
> > Thank you all, for your patience. ❤️
> 
> 
> 
> @SergioBenitez  The docs seem to be wrong
> 
> 
> 
> The docs say to do
> 
> <img width="594" alt="image" src="https://github.com/SergioBenitez/Rocket/assets/20921177/cdff2000-6d7e-4d87-9168-9f16b4b89f22">
> 
> ```
> 
> [dependencies]
> 
> ws = { package = "rocket_ws", version ="=0.1.0-rc.3" }
> 
> ```
> 
> but the crate is not yet in crates.io so you need to do
> 
> https://github.com/dani-garcia/vaultwarden/blob/bc26bfa589c007da9b9be37e1172060f38a948b9/Cargo.toml#L60
> 
> ```
> 
> rocket_ws = { git = 'https://github.com/SergioBenitez/Rocket', rev = "ce441b5f46fdf5cd99cb32b8b8638835e4c2a5fa" } # v0.5 branch
> 
> ```

We always write docs on master as if they're part of the release master last targeted. Otherwise we'd need to update the docs with a commit rev every time we push them, which is simply untenable. The docs are there as a reference for those using master, so the assumption is they understand how to depend on Rocket and its dependencies from master properly.
